internal/eval/eval.go

```diff
@@ -328,6 +330,153 @@ func evalGETDEL(args []string, store *dstore.Store) []byte {
 	return clientio.Encode(obj.Value, false)
 }
 
+// evalJSONDEBUG reports value's memmory usage in bytes
+// Returns arity error if subcommand is missing
+// Supports only two subcommand as of now - HELP and MEMORY
+func evalJSONDebug(args []string, store *dstore.Store) []byte {
+	if len(args) < 1 {
+		return diceerrors.NewErrArity("JSON.DEBUG")
+	}
+	subcommand := strings.ToUpper(args[0])
+	switch subcommand {
+	case "HELP":
+		return evalJSONDebugHelp()
+	case "MEMORY":
+		return evalJSONDebugMemory(args[1:], store)
+	default:
+		return diceerrors.NewErrWithFormattedMessage("unknown subcommand - try `JSON.DEBUG HELP`")
+	}
+}
+
+// evalJSONDebugHelp implements HELP subcommand for evalJSONDebug
+// It returns help text
+// It ignore any other args
+func evalJSONDebugHelp() []byte {
+	memoryText := "MEMORY <key> [path] - reports memory usage"
+	helpText := "HELP                - this message"
+	message := []string{memoryText, helpText}
+	return clientio.Encode(message, false)
+}
+
+// evalJSONDebugMemory implements MEMORY subcommand for evalJSONDebug
+// It returns value's memory usage in bytes
+func evalJSONDebugMemory(args []string, store *dstore.Store) []byte {
+	if len(args) < 1 {
+		return diceerrors.NewErrArity("JSON.DEBUG")
+	}
+	key := args[0]
+
+	// default path is root if not specified
+	path := defaultRootPath
+	if len(args) > 1 {
+		path = args[1] // any more args are ignored for this command altogether
+	}
+
+	obj := store.Get(key)
+	if obj == nil {
+		return clientio.RespZero
+	}
+
+	// check if the object is a valid JSON
+	errWithMessage := object.AssertTypeAndEncoding(obj.TypeEncoding, object.ObjTypeJSON, object.ObjEncodingJSON)
+	if errWithMessage != nil {
+		return errWithMessage
+	}
+
+	// handle root path
+	if path == defaultRootPath {
+		jsonData := obj.Value
+
+		// memory used by json data
+		size := calculateMemory(jsonData)
+		if size == -1 {
+			return diceerrors.NewErrWithMessage("unknown type")
+		}
+
+		// add memory used by storage object
+		size += int(unsafe.Sizeof(obj)) + calculateMemory(obj.LastAccessedAt) + calculateMemory(obj.TypeEncoding)
+
+		return clientio.Encode(size, false)
+	}
+
+	// handle nested paths
+	var results []any = []any{}
+	if path != defaultRootPath {
+
+		// check if path is valid
+		expr, err := jp.ParseString(path)
+		if err != nil {
+			return diceerrors.NewErrWithMessage("invalid JSON path")
+		}
+
+		results = expr.Get(obj.Value)
+
+		// handle error cases
+		if len(results) == 0 {
+			// handle out of bound index path for array json type
+			isArray := utils.IsArray(obj.Value)
+			if isArray {
+				arr := obj.Value.([]any)
+				reg := regexp.MustCompile(`^\$(\.|)\[(\d+|\*)\]`)
+				matches := reg.FindStringSubmatch(path)
+
+				if len(matches) == 3 {
+					index, err := strconv.Atoi(matches[2])
+					if err != nil {
+						return diceerrors.NewErrWithMessage("unable to extract index")
+					}
+					if index >= len(arr) {
+						return clientio.RespEmptyArray
+					}
+				}
+			}
```

Comment:
Can you help me understand what's happening here? Maybe add a few comments here too.

Reply:
Also, we should avoid using regex in general since it is slow to compile and evaluate. However, if there is no better way, we can live with it.

Reply:
@JyotinderSingh - this handles an edge case where memory is requested for out of bound index in an array json. Without this bit of code, program would return an error `Err Path '$.[4]' does not exist` for an array of length 3. Redis returns an empty array.

So this bit of code matches if arg is $[4] type, extracts value 4 and compares it with length to handle this case.

I will add comments here.